x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java

```diff
@@ -73,7 +77,18 @@ private void preFlight(final Operation operation) {
     @Override
     protected InternalEngine.DeletionStrategy deletionStrategyForOperation(final Delete delete) throws IOException {
         preFlight(delete);
-        return planDeletionAsNonPrimary(delete);
+        if (delete.origin() == Operation.Origin.PRIMARY && hasBeenProcessedBefore(delete)) {
+            final AlreadyProcessedFollowingEngineException error = new AlreadyProcessedFollowingEngineException(shardId, delete.seqNo());
+            return DeletionStrategy.skipDueToVersionConflict(error, delete.version(), delete.primaryTerm(), false);
+        } else {
+            return planDeletionAsNonPrimary(delete);
+        }
+    }
+
+    @Override
+    public NoOpResult noOp(NoOp noOp) {
+        // TODO: Make sure we process NoOp once.
```

Comment:
why can't we do this now in this PR in the same way?

Reply:
This is because NoOps don't have _id and they are processed without the _id lock. I am not sure if we need to introduce a fake _id (for locking purpose) for Noops.  Thus, I prefer to make it in a separate PR so we can see it more clear.

Reply:
+1